sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -297,6 +364,39 @@ public DataSourceConfiguration withConnectionProperties(
       return builder().setConnectionProperties(connectionProperties).build();
     }
 
+    /**
+     * Allows to tweak the {@link PoolingDataSource} created internally by {@link JdbcIO}.
+     */
+    public DataSourceConfiguration withPoolConfiguration(int maxTotal,
```

Comment:
Please do not proliferate configuration parameters. This is a bad practice that's come up many types before, we should allow people to set only the parameters which we absolutely have no way of setting to a reasonable value ourselves. Configurability is not a goal in itself, especially in Beam, where the runner has a lot of discretion in how to run the job, so the user has no way of controlling the impact of many parameters.

In this particular case, moreover, almost all of these configuration parameters are unnecessary and misleading - we are creating a pool per DoFn instance, and DoFn instances are accessed sequentially - so we need a pool with capacity for exactly 1 connection, that will expire and be closed if not used for a certain amount of time, that is all.

Reply:
OK, I wanted to give complete control to users. I think, only the validationQuery is required IMHO.

Reply:
Yeah I understand the intention, but these parameters are intended to be tuned according to the user's particular usage pattern, and here the usage pattern is almost completely outside the user's control - it's hidden behind:

* implementation details of JdbcIO which are allowed to change at any moment without notice (and in fact *are* changing - this PR being an example)
* implementation details of the particular runner, which also can do anything it wants as long as it complies with the Beam model; different runners do different things at different times with different data

So there does not exist any goal that can be reliably accomplished by changing these parameters.

This is covered in https://beam.apache.org/contribute/ptransform-style-guide/#what-parameters-to-expose - I'm happy to expand that section with the argument above, if you think that'd be useful.

validationQuery is not required either. AFAIK since JDBC4 (introduced in 2006, with Java 6), JDBC connections can validate themselves in a driver-specific fashion via connection.isValid(), and dbcp takes advantage of that.